lib/experimental/html/class-wp-html-attribute-sourcer.php

```diff
@@ -0,0 +1,390 @@
+			return '';
+		}
+
+		while ( $tags->next_tag( [ 'tag_closers' => 'visit' ] ) ) {
+			if ( $tags->get_tag() !== $tag_name ) {
+				continue;
+			}
+
+			if ( $tags->is_tag_closer() && $depth === 1 ) {
+				$tags->set_bookmark( 'end' );
+				break;
+			}
+
+			$depth += $tags->is_tag_closer() ? -1 : 1;
+		}
+
+		return $tags->inner_content( 'start', 'end' );
+	}
+
+	/**
+	 * @see https://html.spec.whatwg.org/#elements-2
+	 */
+	public static function is_void_element( $tag_name ) {
+		switch ( $tag_name ) {
+			case 'area':
+			case 'base':
+			case 'br':
+			case 'col':
+			case 'embed':
+			case 'hr':
+			case 'img':
+			case 'input':
+			case 'link':
+			case 'meta':
+			case 'source':
+			case 'track':
+			case 'wbr':
+				return true;
+
+			default:
+				return false;
+		}
+	}
+
+	public static function parse_definition( $definition ) {
+		if ( empty( $definition['source'] ) ) {
+			return 'not-sourced';
+		}
+
+		$source = $definition['source'];
+		if ( 'html' !== $source && 'attribute' !== $source ) {
+			return 'unsupported';
+		}
+
+		if ( 'attribute' === $source && empty( $definition['selector'] ) ) {
+			return null;
+		}
+
+		if ( 'html' === $source && empty( $definition['selector'] ) ) {
+			return 'inner-html';
+		}
+
+		$selector = self::parse_selector( $definition['selector'] );
+		if ( null === $selector ) {
+			return 'unsupported';
+		}
+
+		if ( 'html' === $source ) {
+			return array( 'type' => 'html', 'selector' => $selector );
+		}
+
+		$attribute = self::parse_attribute( $definition['attribute'] );
+		if ( null === $attribute ) {
+			return null;
+		}
+
+		return array( 'type' => 'attribute', 'selector' => $selector, 'attribute' => $attribute );
+	}
+
+	public static function parse_selector( $s, $at = 0 ) {
+		$selectors = explode( ',', $s );
+		if ( count( $selectors ) > 1 ) {
+			$parsed = [];
+
+			foreach ( $selectors as $selector ) {
+				$parsed[] = self::parse_selector( $selector, strspn( $selector, " \r\t\f\n" ) );
+			}
+
+			return $parsed;
+		}
+
+		$type = 'element';
+
+		switch ( $s[ $at ] ) {
+			case '+':
+				// no support for adjacent sibling combinator
+				return null;
+
+			case '>':
+				// no support for child combinator
+				return null;
+
+			case '~':
+				// no support for general sibling combinator
+				return null;
+
+			case ' ':
+				// no support for descendant combinator
+				return null;
+
+			case '[':
+				// no support for attribute
+				return null;
+
+			case ',':
+				// we shouldn't get here because we're exploding at the start
+				// of this function; this is a bug if we're here.
+				return null;
+
+			case ':':
+				// no support for pseudo-selectors
+				return null;
+
+			case '#':
+				$type = 'hash';
+				$at++;
+				break;
+
+			case '.':
+				$type = 'class';
+				$at++;
+				break;
+		}
+
+		// @TODO: Hashes don't have to start with `nmstart` so this might reject valid hash names.
+		$identifier = self::parse_css_identifier( $s, $at );
```

Comment:
Looks like it only consumes the first css identifier at the moment, like `figure` in `figure video`.

Reply:
looks right 😉 

though to be fair if you look up at line 275 above it, it's acknowledged that we don't support the descendant selector yet